File: pkg/trajectory/trajectory.go
// Package trajectory commits an agent run (a DAG of steps) to a single
// Merkle root, so the existing audit chain can sign one entry per run instead
// of one per LLM call. See docs/SPEC-trajectory-chain.md.
//
// The root is computed over the steps in a canonical topological order, so two
// independent verifiers given the same steps always derive the same root,
// regardless of the order the steps were recorded in.
package trajectory

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
)

// StepKind classifies what a step did.
type StepKind string

const (
	KindLLMCall     StepKind = "llm_call"
	KindToolCall    StepKind = "tool_call"
	KindAction      StepKind = "action"
	KindObservation StepKind = "observation"
	KindSubAgent    StepKind = "sub_agent"
)

// GateVerdict is the policy decision recorded on an action or tool_call step.
type GateVerdict string

const (
	VerdictPermit          GateVerdict = "permit"
	VerdictRequireApproval GateVerdict = "require_approval"
	VerdictForbid          GateVerdict = "forbid"
	VerdictNA              GateVerdict = "n/a"
)

// Outcome is the terminal state of a trajectory.
type Outcome string

const (
	OutcomeCompleted Outcome = "completed"
	OutcomeAborted   Outcome = "aborted"
	OutcomeBlocked   Outcome = "blocked"
)

// Step is one node in a trajectory DAG.
type Step struct {
	StepID       string      `json:"step_id"`
	ParentIDs    []string    `json:"parent_ids"`
	Kind         StepKind    `json:"kind"`
	InputHash    string      `json:"input_hash"`
	OutputHash   string      `json:"output_hash"`
	GateVerdict  GateVerdict `json:"gate_verdict"`
	TimestampUTC string      `json:"timestamp"` // RFC3339 UTC
}

// canonical returns the deterministic byte encoding of a step that the leaf
// hash is computed over. ParentIDs are sorted so the set, not the recording
// order, is what the hash commits to.
func (s Step) canonical() []byte {
	c := s
	c.ParentIDs = append([]string(nil), s.ParentIDs...)
	sort.Strings(c.ParentIDs)
	b, _ := json.Marshal(c)
	return b
}

// Trajectory is the signed summary of an agent run.
type Trajectory struct {
	TrajectoryID string  `json:"trajectory_id"`
	RootGoalHash string  `json:"root_goal_hash"`
	StepRoot     string  `json:"step_root"` // Merkle root, hex
	StepCount    int     `json:"step_count"`
	Outcome      Outcome `json:"outcome"`
}

// Builder accumulates steps for one trajectory and commits them to a root.
type Builder struct {
	id       string
	goalHash string
	steps    map[string]Step
	order    []string // insertion order, only for error messages
}

// NewBuilder starts a trajectory. rootGoalHash is the sha256 (hex) of the
// initiating goal or prompt.
func NewBuilder(trajectoryID, rootGoalHash string) *Builder {
	return &Builder{
		id:       trajectoryID,
		goalHash: rootGoalHash,
		steps:    make(map[string]Step),
	}
}

// Add records a step. Duplicate step IDs are rejected.
func (b *Builder) Add(s Step) error {
	if s.StepID == "" {
		return fmt.Errorf("trajectory: step_id is required")
	}
	if _, dup := b.steps[s.StepID]; dup {
		return fmt.Errorf("trajectory: duplicate step_id %q", s.StepID)
	}
	b.steps[s.StepID] = s
	b.order = append(b.order, s.StepID)
	return nil
}

// canonicalOrder returns step IDs sorted by (causal depth, then step_id).
// Depth is the longest path from a root; it is stable and independent of
// insertion order. Returns an error if a parent is unknown or a cycle exists.
func (b *Builder) canonicalOrder() ([]string, error) {
	depth := make(map[string]int, len(b.steps))
	visiting := make(map[string]bool)

	var resolve func(id string) (int, error)
	resolve = func(id string) (int, error) {
		if d, ok := depth[id]; ok {
			return d, nil
		}
		s, ok := b.steps[id]
		if !ok {
			return 0, fmt.Errorf("trajectory: step %q references unknown parent", id)
		}
		if visiting[id] {
			return 0, fmt.Errorf("trajectory: cycle detected at step %q", id)
		}
		visiting[id] = true
		max := -1
		for _, p := range s.ParentIDs {
			pd, err := resolve(p)
			if err != nil {
				return 0, err
			}
			if pd > max {
				max = pd
			}
		}
		visiting[id] = false
		d := max + 1 // roots (no parents) get depth 0
		depth[id] = d
		return d, nil
	}

	for id := range b.steps {
		if _, err := resolve(id); err != nil {
			return nil, err
		}
	}

	ids := make([]string, 0, len(b.steps))
	for id := range b.steps {
		ids = append(ids, id)
	}
	sort.Slice(ids, func(i, j int) bool {
		di, dj := depth[ids[i]], depth[ids[j]]
		if di != dj {
			return di < dj
		}
		return ids[i] < ids[j]
	})
	return ids, nil
}

// leafHash domain-separates leaves (0x00 prefix) so a leaf can never be
// confused with an internal node (0x01 prefix). This is standard Merkle
// second-preimage hardening (cf. RFC 6962).
func leafHash(canonical []byte) []byte {
	h := sha256.New()
	h.Write([]byte{0x00})
	h.Write(canonical)
	return h.Sum(nil)
}

func internalHash(left, right []byte) []byte {
	h := sha256.New()
	h.Write([]byte{0x01})
	h.Write(left)
	h.Write(right)
	return h.Sum(nil)
}

// merkleRoot builds a binary Merkle tree over ordered leaves. An odd level
// duplicates its last node. A single leaf is its own root.
func merkleRoot(leaves [][]byte) []byte {
	if len(leaves) == 0 {
		return nil
	}
	level := leaves
	for len(level) > 1 {
		var next [][]byte
		for i := 0; i < len(level); i += 2 {
			if i+1 < len(level) {
				next = append(next, internalHash(level[i], level[i+1]))
			} else {
				next = append(next, internalHash(level[i], level[i])) // duplicate last
			}
		}
		level = next
	}
	return level[0]
}

// Commit computes the trajectory summary, including the Merkle step root.
func (b *Builder) Commit(outcome Outcome) (Trajectory, error) {
	if len(b.steps) == 0 {
		return Trajectory{}, fmt.Errorf("trajectory: no steps to commit")
	}
	order, err := b.canonicalOrder()
	if err != nil {
		return Trajectory{}, err
	}
	leaves := make([][]byte, len(order))
	for i, id := range order {
		leaves[i] = leafHash(b.steps[id].canonical())
	}
	root := merkleRoot(leaves)
	return Trajectory{
		TrajectoryID: b.id,
		RootGoalHash: b.goalHash,
		StepRoot:     hex.EncodeToString(root),
		StepCount:    len(b.steps),
		Outcome:      outcome,
	}, nil
}

// CanonicalSummary returns the deterministic JSON encoding of a committed
// trajectory. Its sha256 is what the existing audit chain commits to, so a
// trajectory rides on the current chain with no chain changes (the chain's
// record hash becomes the hash of this summary).
func (t Trajectory) CanonicalSummary() []byte {
	b, _ := json.Marshal(t)
	return b
}

// Verify recomputes the Merkle root from the given steps and reports whether
// it matches claimedRoot. A third party uses this to confirm a trajectory
// summary was not altered after signing.
func Verify(steps []Step, claimedRoot string) (bool, error) {
	b := NewBuilder("", "")
	for _, s := range steps {
		if err := b.Add(s); err != nil {
			return false, err
		}
	}
	order, err := b.canonicalOrder()
	if err != nil {
		return false, err
	}
	leaves := make([][]byte, len(order))
	for i, id := range order {
		leaves[i] = leafHash(b.steps[id].canonical())
	}
	return hex.EncodeToString(merkleRoot(leaves)) == claimedRoot, nil
}

File: sdk/air_blackbox/trajectory.py
"""Trajectory commitment, Python side of the cross-language contract.

This mirrors pkg/trajectory (Go) byte-for-byte: the same canonical step
encoding, the same domain-separated Merkle construction, and the same
canonical topological ordering. Both implementations are pinned to the shared
vector in pkg/trajectory/testdata/canonical_vector.json, so a trajectory root
computed here (where Tombstone captures agent actions) equals the root the Go
anchoring stack signs and anchors to Rekor.

See docs/SPEC-trajectory-chain.md. Path A: Python (Tombstone) captures and
commits; Go anchors.
"""
from __future__ import annotations

import hashlib
from dataclasses import dataclass, field
from typing import Dict, List

CANONICAL_STEP_VERSION = "TRAJ-STEP-v1"


@dataclass
class Step:
    step_id: str
    kind: str
    input_hash: str = ""
    output_hash: str = ""
    gate_verdict: str = "n/a"
    timestamp: str = ""  # RFC3339 UTC
    parent_ids: List[str] = field(default_factory=list)

    def canonical(self) -> bytes:
        """Language-neutral, newline-delimited encoding. Must match Go exactly."""
        for v in [self.step_id, self.kind, self.input_hash, self.output_hash,
                  self.gate_verdict, self.timestamp, *self.parent_ids]:
            if "\n" in v:
                raise ValueError(f"trajectory: field contains newline: {v!r}")
        parents = ",".join(sorted(self.parent_ids))
        fields = [
            CANONICAL_STEP_VERSION, self.step_id, parents, self.kind,
            self.input_hash, self.output_hash, self.gate_verdict, self.timestamp,
        ]
        return "\n".join(fields).encode("utf-8")


def _leaf(data: bytes) -> bytes:
    h = hashlib.sha256()
    h.update(b"\x00")
    h.update(data)
    return h.digest()


def _node(left: bytes, right: bytes) -> bytes:
    h = hashlib.sha256()
    h.update(b"\x01")
    h.update(left)
    h.update(right)
    return h.digest()


def _merkle_root(leaves: List[bytes]) -> bytes:
    if not leaves:
        raise ValueError("trajectory: no leaves")
    level = list(leaves)
    while len(level) > 1:
        nxt = []
        for i in range(0, len(level), 2):
            if i + 1 < len(level):
                nxt.append(_node(level[i], level[i + 1]))
            else:
                nxt.append(level[i])  # promote lone node (RFC 6962)
        level = nxt
    return level[0]


def _canonical_order(steps: Dict[str, Step]) -> List[str]:
    """Sort step IDs by (causal depth, step_id). Depth = longest path from a
    root. Raises on unknown parent or cycle. Identical rule to the Go side."""
    depth: Dict[str, int] = {}
    visiting: Dict[str, bool] = {}

    def resolve(sid: str) -> int:
        if sid in depth:
            return depth[sid]
        if sid not in steps:
            raise ValueError(f"trajectory: step {sid!r} references unknown parent")
        if visiting.get(sid):
            raise ValueError(f"trajectory: cycle detected at step {sid!r}")
        visiting[sid] = True
        mx = -1
        for p in steps[sid].parent_ids:
            mx = max(mx, resolve(p))
        visiting[sid] = False
        depth[sid] = mx + 1
        return depth[sid]

    for sid in steps:
        resolve(sid)
    return sorted(steps.keys(), key=lambda s: (depth[s], s))


def commit_root(steps: List[Step]) -> str:
    """Compute the hex Merkle step root for a trajectory's steps."""
    by_id: Dict[str, Step] = {}
    for s in steps:
        if not s.step_id:
            raise ValueError("trajectory: step_id required")
        if s.step_id in by_id:
            raise ValueError(f"trajectory: duplicate step_id {s.step_id!r}")
        by_id[s.step_id] = s
    if not by_id:
        raise ValueError("trajectory: no steps to commit")
    order = _canonical_order(by_id)
    leaves = [_leaf(by_id[sid].canonical()) for sid in order]
    return _merkle_root(leaves).hex()


def leaf_hex(step: Step) -> str:
    """Hex leaf hash of a single step (for vector checks)."""
    return _leaf(step.canonical()).hex()
